clean up error paths in modules/resource-hwloc

garlick opened the pull request:
Replace resource-hwloc calls to `FLUX_CHECK_INT()` and `FLUX_CHECK_PTR()` with error paths that return RPC error responses or cause `mod_main()` to exit, with copious logging.  A bit of refactoring for clarity as well.

Fixes issue #536

grondo commented:
Nice work! There are no longer any users of `FLUX_CHECK_*` macros. Should we remove the macros as part of this PR to avoid encourage use of calls that exit() from modules?

grondo commented:
I also notice some other modules that call `msg`/`err`/`_exit()` which might be cleaned up:
```
 git grep -E --line-number '(msg|err)_exit' .                                                                                              
barrier/barrier.c:78:            err_exit ("flux_get_rank");
connector-local/local.c:115:            err_exit ("flux_get_reactor");
kvs/kvs.c:816:            msg_exit ("%s: unexpected DIRVAL: path=%s name=%s: dirent=%s ",
kvs/kvs.c:823:            msg_exit ("%s: unknown dirent type: path=%s name=%s: dirent=%s ",
kvs/kvs.c:922:            msg_exit ("%s: corrupt internal storage", __FUNCTION__);
live/live.c:178:            err_exit ("flux_get_rank");
live/live.c:189:            err_exit ("flux_reduce_create");
pymod/py_mod.c:94:        msg_exit ("optparse_add_option_table");
pymod/py_mod.c:96:        msg_exit ("optparse_set usage");
wreck/job.c:146:        err_exit ("kvs_get_dir (id=%lu)", id);
wreck/wrexec.c:86:            err_exit ("flux_get_rank");
wreck/wrexec.c:88:            err_exit ("flux_attr_get local-uri");
wreck/wrexec.c:160:        err_exit ("fork");
wreck/wrexec.c:172:        err_exit ("setenv");
wreck/wrexec.c:175:        err_exit ("execvp");
wreck/wrexec.c:192:        err_exit ("fork");
```

Want me to open issue(s) on these? (Is one issue ok?)

garlick commented:
Yes and yes.  Thanks.

On Tuesday, February 2, 2016, Mark Grondona <notifications@github.com>
wrote:

> I also notice some other modules that call msg/err/_exit() which might be
> cleaned up:
>
>  git grep -E --line-number '(msg|err)_exit' .
> barrier/barrier.c:78:            err_exit ("flux_get_rank");
> connector-local/local.c:115:            err_exit ("flux_get_reactor");
> kvs/kvs.c:816:            msg_exit ("%s: unexpected DIRVAL: path=%s name=%s: dirent=%s ",
> kvs/kvs.c:823:            msg_exit ("%s: unknown dirent type: path=%s name=%s: dirent=%s ",
> kvs/kvs.c:922:            msg_exit ("%s: corrupt internal storage", __FUNCTION__);
> live/live.c:178:            err_exit ("flux_get_rank");
> live/live.c:189:            err_exit ("flux_reduce_create");
> pymod/py_mod.c:94:        msg_exit ("optparse_add_option_table");
> pymod/py_mod.c:96:        msg_exit ("optparse_set usage");
> wreck/job.c:146:        err_exit ("kvs_get_dir (id=%lu)", id);
> wreck/wrexec.c:86:            err_exit ("flux_get_rank");
> wreck/wrexec.c:88:            err_exit ("flux_attr_get local-uri");
> wreck/wrexec.c:160:        err_exit ("fork");
> wreck/wrexec.c:172:        err_exit ("setenv");
> wreck/wrexec.c:175:        err_exit ("execvp");
> wreck/wrexec.c:192:        err_exit ("fork");
>
> Want me to open issue(s) on these? (Is one issue ok?)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/539#issuecomment-178957181>
> .
>

grondo commented:
While looking at why no coverage data was being reported for `modules/resource-hwloc/resource.c` I found that (apparently) when you use per-target `_CFLAGS` (perhaps `_LDFLAGS` too), this overrides  `AM_CFLAGS`. This means that, in addition to not compiling with coverage support, `resource-hwloc` is not being built with `-Werror`. Once fixed, I did get an error for `resource.c` that you may want to fix here:
```
$ make
  CC       resource_hwloc_la-resource.lo
resource.c: In function ‘reload_request_cb’:
resource.c:359:9: error: unused variable ‘rc’ [-Werror=unused-variable]
     int rc, errnum = 0;
         ^
cc1: all warnings being treated as errors
make: *** [resource_hwloc_la-resource.lo] Error 1
```

grondo commented:
From [Automake manual](http://www.gnu.org/software/automake/manual/html_node/Program-Variables.html#Program-Variables)
```
AM_CFLAGS
This is the variable the Makefile.am author can use to pass in additional C compiler flags. In some situations, this is not used, in preference to the per-executable (or per-library) _CFLAGS.
```

garlick commented:
Thanks, I fixed the warning.  Shall I fix the CFLAGS here too?

grondo commented:
Yeah, I was going to suggest that:
```diff
diff --git a/src/modules/resource-hwloc/Makefile.am b/src/modules/resource-hwloc/Makefile.am
index fb2295f..b6daa7f 100644
--- a/src/modules/resource-hwloc/Makefile.am
+++ b/src/modules/resource-hwloc/Makefile.am
@@ -15,7 +15,7 @@ AM_CPPFLAGS = \
 fluxmod_LTLIBRARIES = resource-hwloc.la
 
 resource_hwloc_la_SOURCES = resource.c
-resource_hwloc_la_CFLAGS = $(HWLOC_CFLAGS)
+resource_hwloc_la_CFLAGS = $(AM_CFLAGS) $(HWLOC_CFLAGS)
 resource_hwloc_la_LDFLAGS = $(fluxmod_ldflags) -module
 resource_hwloc_la_LIBADD = $(top_builddir)/src/common/libflux-core.la \
                           $(top_builddir)/src/common/libflux-internal.la \

```

grondo commented:
Thanks! Is this one ready to merge?

garlick commented:
OK, I think this might be ready to go if it passes travis.

The following are included here though they are a bit off topic for this PR:
* fixed non-oom exits in live module
* modernized api usage in live module
* fixed non-oom exits in barrier module
* tweaked `git describe` options in AC_INIT

I skipped doing anything with the kvs code because I have other invasive changes pending on it.

garlick commented:
Just restarted travis - it gcc had failed on that same jstat test that's failed intermittently before, and clang had hung but I was in a hurry and didn't dig into it.

dongahn commented:
It is still on my todo list to look at that intermittent jstat test failure. Once I successfully build the docker image on my virtualBox, my hope is to see this nondeterministic problem more frequently to be able to debug.  

Sorry, I have too many burning issues from other projects/tasks on my plate at this point, which keeps me from spending quality time on Flux.

grondo commented:
OK, looks good!
